File: ios/Ribbon/DesignSystem/WaveMark.swift
import SwiftUI

// The Wave — the mark (W6 study, from /mark). Two tapered ribbons falling
// from a common edge, crossing low, each ending in a swallowtail. In the
// app it is: the only branded moment (onboarding), and the way out of the
// book (S02/S03).
//
// Occlusion, not transparency: the front ribbon knocks a hard-edged gap out
// of the back one. Drawn natively from the generated geometry so it is
// crisp at 20 pt and at 200 pt.

private func wavePath(_ segments: [WaveSegment], in rect: CGRect) -> Path {
    let scale = min(rect.width, rect.height) / WaveGeometry.gridSize
    let dx = rect.midX - WaveGeometry.gridSize * scale / 2
    let dy = rect.midY - WaveGeometry.gridSize * scale / 2
    func point(_ x: CGFloat, _ y: CGFloat) -> CGPoint {
        CGPoint(x: x * scale + dx, y: y * scale + dy)
    }
    var path = Path()
    for segment in segments {
        switch segment {
        case let .move(x, y): path.move(to: point(x, y))
        case let .line(x, y): path.addLine(to: point(x, y))
        case let .quad(cx, cy, x, y): path.addQuadCurve(to: point(x, y), control: point(cx, cy))
        case .close: path.closeSubpath()
        }
    }
    return path
}

struct WaveMark: View {
    var color: Color = Palette.chartreuse

    var body: some View {
        Canvas { context, size in
            let rect = CGRect(origin: .zero, size: size)
            let scale = min(size.width, size.height) / WaveGeometry.gridSize
            let back = wavePath(WaveGeometry.back, in: rect)
            let front = wavePath(WaveGeometry.front, in: rect)

            // Back ribbon, then the knockout: erase a stroke of the front
            // ribbon's outline from what's drawn so far, then lay the front
            // ribbon in. The gap is the ground showing through — a hard
            // edge, no opacity blending.
            context.drawLayer { layer in
                layer.fill(back, with: .color(color))
                layer.blendMode = .destinationOut
                layer.stroke(
                    front,
                    with: .color(.black),
                    style: StrokeStyle(
                        lineWidth: WaveGeometry.knockoutWidth * scale,
                        lineJoin: .miter, miterLimit: 6))
            }
            context.fill(front, with: .color(color))
        }
        .accessibilityHidden(true)
    }
}

#Preview("The Wave") {
    VStack(spacing: 40) {
        WaveMark().frame(width: 200, height: 200)
        WaveMark(color: Palette.text.opacity(0.6)).frame(width: 20, height: 20)
    }
    .frame(maxWidth: .infinity, maxHeight: .infinity)
    .background(Palette.ground)
}

File: android/app/src/main/kotlin/app/readribbon/design/WaveMark.kt
package app.readribbon.design

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.scale
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

// The Wave (W6) — the mark, drawn natively at any size from the geometry
// tools/make_assets.py emits. It is the way out of the book everywhere it
// appears, and the closing drag is always duplicated by it (§11, motor).

internal fun List<WaveSegment>.toPath(): Path {
    val path = Path()
    for (segment in this) {
        when (segment) {
            is WaveSegment.Move -> path.moveTo(segment.x, segment.y)
            is WaveSegment.Line -> path.lineTo(segment.x, segment.y)
            is WaveSegment.Quad -> path.quadraticTo(segment.cx, segment.cy, segment.x, segment.y)
            WaveSegment.Close -> path.close()
        }
    }
    return path
}

/**
 * Draws the mark into the current [DrawScope], filling a square of [side]
 * pixels from the origin.
 *
 * The knockout is occlusion, not transparency: the back ribbon is filled,
 * then the front ribbon's outline is stroked in the ground colour, then the
 * front ribbon is filled over the inner half of its own stroke. One ribbon
 * passes behind the other with a hard edge — which is why [ground] has to be
 * whatever this mark is actually sitting on, and why the mark is never drawn
 * over a photograph.
 */
fun DrawScope.drawWave(
    side: Float,
    front: Path,
    back: Path,
    tint: Color = Palette.chartreuse,
    ground: Color = Palette.ground,
) {
    scale(scale = side / WaveGeometry.GRID_SIZE, pivot = Offset.Zero) {
        drawPath(back, tint)
        drawPath(
            path = front,
            color = ground,
            style = Stroke(
                width = WaveGeometry.KNOCKOUT_WIDTH,
                join = StrokeJoin.Miter,
                miter = 6f,
            ),
        )
        drawPath(front, tint)
    }
}

/**
 * The mark as a view. The mark's own 64-unit grid is scaled into [size].
 *
 * Sizing is the drawn size only. Every quiet control and both Waves out of
 * the book meet the 44 dp touch minimum (deviation 12) — that is the
 * caller's business, because the mark is usually drawn smaller than the
 * target that carries it.
 */
@Composable
fun WaveMark(
    modifier: Modifier = Modifier,
    size: Dp = 28.dp,
    tint: Color = Palette.chartreuse,
    ground: Color = Palette.ground,
) {
    val front = remember { WaveGeometry.front.toPath() }
    val back = remember { WaveGeometry.back.toPath() }
    Canvas(modifier = modifier.size(size)) {
        drawWave(
            side = minOf(this.size.width, this.size.height),
            front = front,
            back = back,
            tint = tint,
            ground = ground,
        )
    }
}
